text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,776 @@
+# Programmatic Toolkit
+
+* **Original Author(s):**: @mrgrain
+* **Tracking Issue**: #300
+* **API Bar Raiser**: @rix0rrr
+
+The [Programmatic Toolkit] allows integrators to build lifecycle management solutions for CDK applications.
+It provides them with full control over all output and allows to interweave actions with custom steps.
+
+## Working Backwards
+
+### README
+
+[Programmatic Toolkit] is a library to run CDK actions programmatically.
+The package provides implementations for the following actions:
+
+- `bootstrap`
+- `list`
+- `synth`
+- `deploy`
+- ...
+
+#### Self-contained example
+
+```ts
+// Create a [Programmatic Toolkit]
+await using cdk = new Toolkit({
+  ioHost: new CloudWatchLogsIoHost('my-log-stream'), // optional, e.g. an IoHost to write any output to CloudWatch Logs
+});
+
+// Define the CloudExecutable
+const cx = CloudExecutable.fromCloudAssemblyProducer(async (context) => {
+  const app = new cdk.App({ context });
+  const stack = new cdk.Stack(app);
+  return app.synth();
+});
+
+// Synth and store a cached CloudExecutable, this won't run synth again
+const cxCache = await cdk.synth(cx);
+
+// Deploy our pre-synth'ed app
+const deployment = await cdk.deploy(cxCache, { stacks: ['MyTestStack'] });
+
+// Returns a list of ARNs of the deployment
+return deployment.listResources().filter(r => r.arn);
```

Comment:
Does this method also allow access to resource attributes like `AWS::Lambda::Url`'s `FunctionUrl`? I could see this being very helpful for automated testing.

Reply:
Unfortunately I don't think that's easily possible as CFN stack APIs [don't include this information](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_ListStackResources.html). You'd need to use a Stack output or make the request to the service API manually. But at least that should be much easier.

Reply:
Stack outputs would work fine, but how would that look work under this proposal? Is there a separate `deployment.listOutputs()` or something like that?